synedrion/src/uint/secret_signed.rs

```diff
@@ -0,0 +1,710 @@
+            self.bound,
+        )
+        .expect("the absolute value is within the same bound")
+    }
+
+    /// Creates a [`SignedSecret`] from an unsigned value, treating it as if it encodes the sign as two's complement.
+    ///
+    /// Returns `None` if the requested bound is too large, or if `abs(value)` is actually larger than the bound.
+    pub fn new_from_unsigned(value: Secret<T>, bound: u32) -> Option<Self> {
+        let is_negative = Choice::from(value.expose_secret().bit_vartime(T::BITS - 1) as u8);
+        let abs = Secret::<T>::conditional_select(&value, &value.wrapping_neg(), is_negative);
+        if bound >= T::BITS || abs.expose_secret().bits() > bound {
+            return None;
+        }
+        Some(Self::new_from_unsigned_unchecked(value, bound))
+    }
+
+    /// Creates a [`SignedSecret`] from an unsigned value, treating it as if it is the absolute value.
+    /// If `is_negative` is truthy, crates a negative [`SignedSecret`] with the given absolute value.
+    ///
+    /// Returns `None` if the bound is too large, or if `abs_value` is actually larger than the bound.
+    fn new_from_abs(abs_value: Secret<T>, bound: u32, is_negative: Choice) -> Option<Self> {
+        Self::new_from_unsigned(
+            Secret::<T>::conditional_select(&abs_value, &abs_value.wrapping_neg(), is_negative),
+            bound,
+        )
+    }
+
+    #[cfg(test)]
+    fn test_new_from_abs(abs_value: T, bound: u32, is_negative: bool) -> Option<Self> {
+        Self::new_from_abs(Secret::init_with(|| abs_value), bound, Choice::from(is_negative as u8))
+    }
+
+    #[cfg(test)]
+    fn test_new_from_unsigned(abs_value: T, bound: u32) -> Option<Self> {
+        Self::new_from_unsigned(Secret::init_with(|| abs_value), bound)
+    }
+
+    /// Returns a truthy `Choice` if the absolute value is within the bit bound `bound`.
+    fn in_bound(&self, bound: u32) -> Choice {
+        let abs = self.abs();
+        let mask = T::one().wrapping_neg().wrapping_shl_vartime(bound);
+        let masked = abs & mask;
+        masked.is_zero()
+    }
+
+    /// Asserts that the absolute value is within the bit bound `bound`.
+    /// If that is the case, returns the value with the bound set to it.
+    pub fn ensure_bound(&self, bound: u32) -> CtOption<Self> {
+        let value = Self {
+            value: self.value.clone(),
+            bound,
+        };
+        CtOption::new(value, self.in_bound(bound))
+    }
+
+    /// Asserts that the value is within the interval the paper denotes as $\pm 2^exp$.
+    /// Panics if it is not the case.
+    ///
+    /// That is, the value must be within $[-2^{exp}, 2^{exp}]$
+    /// (See Section 2).
+    pub fn assert_exp_range(&self, exp: u32) {
+        let in_bound = self.in_bound(exp);
+        // Have to check for the ends of the range too
+        let is_end = self.abs().expose_secret().ct_eq(&(T::one() << exp));
+        assert!(bool::from(in_bound | is_end), "out of bounds $\\pm 2^{exp}$",)
+    }
```

Comment:
I think that we should strive to avoid using asserts in non-test code, and instead prefer `Result` or `Option`. Doing this consistently is a lot of work and I am fine doing it bit by bit, but perhaps we can avoid adding *new* asserting code and instead return `Result` here?

Reply:
Yes, I thought of that too, we can make all the proof constructors (where this method is used) return `Result<..., LocalError>`, so they can fail.

Reply:
Not sure how often this is called or how important it is for it to be CT, but I tinkered a bit with making it faster (made it return `bool` for convenience):

```rust
    pub fn assert_exp_range_fast(&self, exp: u32) -> bool {
        let abs = self.abs();
        let mask = T::one().wrapping_neg().wrapping_shl_vartime(exp);
        if abs.expose_secret() == &(mask.not() + T::one()) {
            return true;
        }
        let masked = abs & mask;
        masked.expose_secret() == &T::zero()
    }
```

Benchmark timings for `U2048`:

```
Bounds check/assert_exp_range, small exponent
                        time:   [420.31 ns 425.26 ns 431.98 ns]
                        change: [-1.3407% +0.3532% +2.1482%] (p = 0.72 > 0.05)
                        No change in performance detected.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe
Bounds check/assert_exp_range_fast, small exponent
                        time:   [266.69 ns 271.08 ns 277.42 ns]
                        change: [-0.9114% +0.2448% +1.4652%] (p = 0.71 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) high mild
  6 (6.00%) high severe
Bounds check/assert_exp_range, big exponent
                        time:   [713.33 ns 769.78 ns 821.11 ns]
                        change: [-8.6114% +0.4042% +10.744%] (p = 0.94 > 0.05)
                        No change in performance detected.
Bounds check/assert_exp_range_fast, big exponent
                        time:   [383.40 ns 421.51 ns 462.13 ns]
                        change: [-6.4458% +2.8628% +12.866%] (p = 0.55 > 0.05)
                        No change in performance detected.
```

Reply:
> Not sure how often this is called or how important it is for it to be CT

A bunch of times during the protocol, but on the order of dozens maybe. So given that it takes <1us, I don't think its performance is too important. It does handle secret data, so I think constant-timeness is appropriate.

Reply:
Fair enough. Do you happen to know if `&` is CT?

Reply:
I hope it is, or a lot of `crypto-bigint` stuff is not CT :) It's used in a lot of places